// @effect/platform/dist/cjs/OpenApi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fromApi = exports.annotations = exports.Version = exports.Transform = exports.Title = exports.Summary = exports.Servers = exports.Override = exports.License = exports.Identifier = exports.Format = exports.ExternalDocs = exports.Exclude = exports.Description = exports.Deprecated = void 0;
var Context = _interopRequireWildcard(require("effect/Context"));
var _Function = require("effect/Function");
var _GlobalValue = require("effect/GlobalValue");
var Option = _interopRequireWildcard(require("effect/Option"));
var Schema = _interopRequireWildcard(require("effect/Schema"));
var HttpApi = _interopRequireWildcard(require("./HttpApi.js"));
var HttpApiMiddleware = _interopRequireWildcard(require("./HttpApiMiddleware.js"));
var HttpApiSchema = _interopRequireWildcard(require("./HttpApiSchema.js"));
var HttpMethod = _interopRequireWildcard(require("./HttpMethod.js"));
var JsonSchema = _interopRequireWildcard(require("./OpenApiJsonSchema.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category annotations
 */
class Identifier extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Identifier")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Identifier = Identifier;
class Title extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Title")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Title = Title;
class Version extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Version")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Version = Version;
class Description extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Description")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Description = Description;
class License extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/License")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.License = License;
class ExternalDocs extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/ExternalDocs")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.ExternalDocs = ExternalDocs;
class Servers extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Servers")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Servers = Servers;
class Format extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Format")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Format = Format;
class Summary extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Summary")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Summary = Summary;
class Deprecated extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Deprecated")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Deprecated = Deprecated;
class Override extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Override")() {}
/**
 * @since 1.0.0
 * @category annotations
 */
exports.Override = Override;
class Exclude extends /*#__PURE__*/Context.Reference()("@effect/platform/OpenApi/Exclude", {
  defaultValue: _Function.constFalse
}) {}
/**
 * Transforms the generated OpenAPI specification
 * @since 1.0.0
 * @category annotations
 */
exports.Exclude = Exclude;
class Transform extends /*#__PURE__*/Context.Tag("@effect/platform/OpenApi/Transform")() {}
exports.Transform = Transform;
const contextPartial = tags => {
  const entries = Object.entries(tags);
  return options => {
    let context = Context.empty();
    for (const [key, tag] of entries) {
      if (options[key] !== undefined) {
        context = Context.add(context, tag, options[key]);
      }
    }
    return context;
  };
};
/**
 * @since 1.0.0
 * @category annotations
 */
const annotations = exports.annotations = /*#__PURE__*/contextPartial({
  identifier: Identifier,
  title: Title,
  version: Version,
  description: Description,
  license: License,
  summary: Summary,
  externalDocs: ExternalDocs,
  servers: Servers,
  format: Format,
  override: Override,
  exclude: Exclude,
  transform: Transform
});
const apiCache = /*#__PURE__*/(0, _GlobalValue.globalValue)("@effect/platform/OpenApi/apiCache", () => new WeakMap());
/**
 * @category constructors
 * @since 1.0.0
 */
const fromApi = self => {
  if (apiCache.has(self)) {
    return apiCache.get(self);
  }
  const api = self;
  const jsonSchemaDefs = {};
  let spec = {
    openapi: "3.1.0",
    info: {
      title: Context.getOrElse(api.annotations, Title, () => "Api"),
      version: Context.getOrElse(api.annotations, Version, () => "0.0.1")
    },
    paths: {},
    tags: [],
    components: {
      schemas: jsonSchemaDefs,
      securitySchemes: {}
    },
    security: []
  };
  function makeJsonSchemaOrRef(schema) {
    return JsonSchema.makeWithDefs(schema, {
      defs: jsonSchemaDefs
    });
  }
  function registerSecurity(name, security) {
    if (spec.components.securitySchemes[name]) {
      return;
    }
    const scheme = makeSecurityScheme(security);
    spec.components.securitySchemes[name] = scheme;
  }
  Option.map(Context.getOption(api.annotations, HttpApi.AdditionalSchemas), componentSchemas => {
    componentSchemas.forEach(componentSchema => makeJsonSchemaOrRef(componentSchema));
  });
  Option.map(Context.getOption(api.annotations, Description), description => {
    spec.info.description = description;
  });
  Option.map(Context.getOption(api.annotations, License), license => {
    spec.info.license = license;
  });
  Option.map(Context.getOption(api.annotations, Summary), summary => {
    spec.info.summary = summary;
  });
  Option.map(Context.getOption(api.annotations, Servers), servers => {
    spec.servers = servers;
  });
  Option.map(Context.getOption(api.annotations, Override), override => {
    Object.assign(spec, override);
  });
  api.middlewares.forEach(middleware => {
    if (!HttpApiMiddleware.isSecurity(middleware)) {
      return;
    }
    for (const [name, security] of Object.entries(middleware.security)) {
      registerSecurity(name, security);
      spec.security.push({
        [name]: []
      });
    }
  });
  HttpApi.reflect(api, {
    onGroup({
      group
    }) {
      if (Context.get(group.annotations, Exclude)) {
        return;
      }
      let tag = {
        name: Context.getOrElse(group.annotations, Title, () => group.identifier)
      };
      Option.map(Context.getOption(group.annotations, Description), description => {
        tag.description = description;
      });
      Option.map(Context.getOption(group.annotations, ExternalDocs), externalDocs => {
        tag.externalDocs = externalDocs;
      });
      Option.map(Context.getOption(group.annotations, Override), override => {
        Object.assign(tag, override);
      });
      Option.map(Context.getOption(group.annotations, Transform), fn => {
        tag = fn(tag);
      });
      spec.tags.push(tag);
    },
    onEndpoint({
      endpoint,
      errors,
      group,
      mergedAnnotations,
      middleware,
      payloads,
      successes
    }) {
      if (Context.get(mergedAnnotations, Exclude)) {
        return;
      }
      const path = endpoint.path.replace(/:(\w+)[^/]*/g, "{$1}");
      const method = endpoint.method.toLowerCase();
      let op = {
        tags: [Context.getOrElse(group.annotations, Title, () => group.identifier)],
        operationId: Context.getOrElse(endpoint.annotations, Identifier, () => group.topLevel ? endpoint.name : `${group.identifier}.${endpoint.name}`),
        parameters: [],
        security: [],
        responses: {}
      };
      Option.map(Context.getOption(endpoint.annotations, Description), description => {
        op.description = description;
      });
      Option.map(Context.getOption(endpoint.annotations, Summary), summary => {
        op.summary = summary;
      });
      Option.map(Context.getOption(endpoint.annotations, Deprecated), deprecated => {
        op.deprecated = deprecated;
      });
      Option.map(Context.getOption(endpoint.annotations, ExternalDocs), externalDocs => {
        op.externalDocs = externalDocs;
      });
      middleware.forEach(middleware => {
        if (!HttpApiMiddleware.isSecurity(middleware)) {
          return;
        }
        for (const [name, security] of Object.entries(middleware.security)) {
          registerSecurity(name, security);
          op.security.push({
            [name]: []
          });
        }
      });
      if (payloads.size > 0) {
        const content = {};
        payloads.forEach(({
          ast
        }, contentType) => {
          content[contentType] = {
            schema: makeJsonSchemaOrRef(Schema.make(ast))
          };
        });
        op.requestBody = {
          content,
          required: true
        };
      }
      for (const [status, {
        ast,
        description
      }] of successes) {
        if (op.responses[status]) continue;
        op.responses[status] = {
          description: Option.getOrElse(description, () => "Success")
        };
        ast.pipe(Option.filter(ast => !HttpApiSchema.getEmptyDecodeable(ast)), Option.map(ast => {
          op.responses[status].content = {
            "application/json": {
              schema: makeJsonSchemaOrRef(Schema.make(ast))
            }
          };
        }));
      }
      if (Option.isSome(endpoint.pathSchema)) {
        const schema = makeJsonSchemaOrRef(endpoint.pathSchema.value);
        if ("properties" in schema) {
          Object.entries(schema.properties).forEach(([name, jsonSchema]) => {
            op.parameters.push({
              name,
              in: "path",
              schema: jsonSchema,
              required: schema.required.includes(name),
              ...(jsonSchema.description ? {
                description: jsonSchema.description
              } : {})
            });
          });
        }
      }
      if (!HttpMethod.hasBody(endpoint.method) && Option.isSome(endpoint.payloadSchema)) {
        const schema = makeJsonSchemaOrRef(endpoint.payloadSchema.value);
        if ("properties" in schema) {
          Object.entries(schema.properties).forEach(([name, jsonSchema]) => {
            op.parameters.push({
              name,
              in: "query",
              schema: jsonSchema,
              required: schema.required.includes(name),
              ...(jsonSchema.description ? {
                description: jsonSchema.description
              } : {})
            });
          });
        }
      }
      if (Option.isSome(endpoint.headersSchema)) {
        const schema = makeJsonSchemaOrRef(endpoint.headersSchema.value);
        if ("properties" in schema) {
          Object.entries(schema.properties).forEach(([name, jsonSchema]) => {
            op.parameters.push({
              name,
              in: "header",
              schema: jsonSchema,
              required: schema.required.includes(name),
              ...(jsonSchema.description ? {
                description: jsonSchema.description
              } : {})
            });
          });
        }
      }
      if (Option.isSome(endpoint.urlParamsSchema)) {
        const schema = makeJsonSchemaOrRef(endpoint.urlParamsSchema.value);
        if ("properties" in schema) {
          Object.entries(schema.properties).forEach(([name, jsonSchema]) => {
            op.parameters.push({
              name,
              in: "query",
              schema: jsonSchema,
              required: schema.required.includes(name),
              ...(jsonSchema.description ? {
                description: jsonSchema.description
              } : {})
            });
          });
        }
      }
      for (const [status, {
        ast,
        description
      }] of errors) {
        if (op.responses[status]) continue;
        op.responses[status] = {
          description: Option.getOrElse(description, () => "Error")
        };
        ast.pipe(Option.filter(ast => !HttpApiSchema.getEmptyDecodeable(ast)), Option.map(ast => {
          op.responses[status].content = {
            "application/json": {
              schema: makeJsonSchemaOrRef(Schema.make(ast))
            }
          };
        }));
      }
      if (!spec.paths[path]) {
        spec.paths[path] = {};
      }
      Option.map(Context.getOption(endpoint.annotations, Override), override => {
        Object.assign(op, override);
      });
      Option.map(Context.getOption(endpoint.annotations, Transform), transformFn => {
        op = transformFn(op);
      });
      spec.paths[path][method] = op;
    }
  });
  Option.map(Context.getOption(api.annotations, Transform), transformFn => {
    spec = transformFn(spec);
  });
  apiCache.set(self, spec);
  return spec;
};
exports.fromApi = fromApi;
const makeSecurityScheme = security => {
  const meta = {};
  Option.map(Context.getOption(security.annotations, Description), description => {
    meta.description = description;
  });
  switch (security._tag) {
    case "Basic":
      {
        return {
          ...meta,
          type: "http",
          scheme: "basic"
        };
      }
    case "Bearer":
      {
        const format = Context.getOption(security.annotations, Format).pipe(Option.map(format => ({
          bearerFormat: format
        })), Option.getOrUndefined);
        return {
          ...meta,
          type: "http",
          scheme: "bearer",
          ...format
        };
      }
    case "ApiKey":
      {
        return {
          ...meta,
          type: "apiKey",
          name: security.key,
          in: security.in
        };
      }
  }
};
